### ISSUE TITLE
[SDK 43] Dev client build fails xcodebuild in expo-modules-core

### ISSUE THREAD

**Issue by sshquack** [expo/expo#14948]
### Summary

Tried the dev clients on a managed expo sdk 43 project according to https://docs.expo.dev/clients/getting-started/

It throws the following errors:
```
> expo run:ios

...
...
› Compiling expo-modules-core Pods/ExpoModulesCore » ConcreteViewProp.swift

❌  (node_modules/expo-modules-core/ios/Swift/Views/ConcreteViewProp.swift:26:53)

  24 |     }
  25 |     guard let value = value as? PropType else {
> 26 |       fatalError("Given value `\(String(describing: value))` cannot be casted to `\(String(describing: PropType.self))`")
     |                                                     ^ variable declared in 'guard' condition is not usable in its body
  27 |     }
  28 |     setter(view, value)
  29 |   }

...
› Compiling expo-modules-core Pods/ExpoModulesCore » ModuleDefinitionComponents.swift

❌  (node_modules/expo-modules-core/ios/Swift/Modules/ModuleDefinitionComponents.swift:196:25)

  194 |  Creates the view manager definition that scopes other view-related definitions.
  195 |  */
> 196 | public func viewManager(@ViewManagerDefinitionBuilder _ closure: @escaping () -> ViewManagerDefinition) -> AnyDefinition {
      |                         ^ unknown attribute 'ViewManagerDefinitionBuilder'
  197 |   return closure()
  198 | }
  199 | 

› 2 error(s), and 1 warning(s)

Failed to build iOS project. "xcodebuild" exited with error code 65.
```

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

managed

### What platform(s) does this occur on?

iOS

### SDK Version (managed workflow only)

43

### Environment

```
> expo diagnostics


  Expo CLI 4.12.10 environment info:
    System:
      OS: macOS 10.15.7
      Shell: 5.7.1 - /bin/zsh
    Binaries:
      Node: 14.17.6 - ~/.n/bin/node
      npm: 8.1.1 - ~/.n/bin/npm
      Watchman: 2021.08.30.00 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.11.0 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: iOS 14.4, DriverKit 20.2, macOS 11.1, tvOS 14.3, watchOS 7.2
    IDEs:
      Android Studio: Arctic Fox 2020.3.1 Patch 3 Arctic Fox 2020.3.1 Patch 3
      Xcode: 12.4/12D4e - /usr/bin/xcodebuild
    npmPackages:
      expo: ~43.0.0 => 43.0.1 
      react: 17.0.1 => 17.0.1 
      react-dom: 17.0.1 => 17.0.1 
      react-native: 0.64.2 => 0.64.2 
      react-native-web: 0.17.1 => 0.17.1 
    npmGlobalPackages:
      expo-cli: 4.12.10
    Expo Workflow: bare
```

### Reproducible demo or steps to reproduce from a blank project

- Expo init `expo init DemoApp01` select "Typescript (with tabs)" template
- Verify all versions in package.json are in sync `expo upgrade 43`
- Run in simulator `react-native run-ios` and it works as expected
- Verify building for simulator works `expo build:ios` and it succeeds as expected
- Add dev client `npm install expo-dev-client` according to https://docs.expo.dev/clients/getting-started/
- Ran `expo run:ios` and it fails xcodebuild
 
See https://gist.github.com/sshquack/1cfdafbfd3a2392a7f8099804f87341c for full error

**Comment by sshquack** [comment]
Thanks for the tips!

@byCedric I tried all of those but still see the same errors. Seems like I need Xcode 12.5

@lukmccall Unfortunately Xcode 12.5 requires macOS Big Sur 11 which in-turn requires a macbook pro 2013+ per [Apple spec](https://support.apple.com/kb/sp833?locale=en_US) I'm on a older 2012 model, so no OS or Xcode upgrades for me 😭

Feel free to close this issue as a duplicate of https://github.com/expo/expo/issues/14856. I won't be trying expo dev clients until I can upgrade my hardware.

**Comment by ZeiRashed** [comment]
Hi all, this issue still repros for me on XCode 13.1. Do we need to be on exactly 12.5?

**Comment by brentvatne** [comment]
sdk 43 requires xcode >= 13 -- xcode 13.1 should work fine though.

@sshquack @ZeiRashed - any chance you could try diffing the result from following the steps of the reproducible example above and the following?

### alternative steps to create dev client project

1. `expo init` - tabs template
2. cd into project
3. `expo install expo-dev-client`
4. `expo prebuild`
5. `expo run:ios` should work for you here

once this is done, compare any of the relevant files with something like https://www.diffchecker.com/ to see what may be different

**Comment by sshquack** [comment]
> * `expo init` - tabs template
> * cd into project
> * `expo install expo-dev-client`
> * `expo prebuild`
> * `expo run:ios` should work for you here

@brentvatne Followed those steps and I consistently get the exact same error as my original error. See full logs here https://gist.github.com/sshquack/3a51edcf81b1171b3d15b03807110fec

> sdk 43 requires xcode >= 13 -- xcode 13.1 should work fine though.

That's the crux of my problem. See my expo diagnostics. My MBP 2012 is only supported until macOS Catalina 10.15 and Xcode 12.4 is the last supported version for 10.15 per [Apple release notes](https://developer.apple.com/documentation/xcode-release-notes/xcode-12_4-release-notes). I would need a MBP 2013 which supports macOS Big Sur 11 which support Xcode 12.5 per [Apple release notes](https://developer.apple.com/documentation/xcode-release-notes/xcode-12_5-release-notes).

**In other words, expo-dev-client transitively requires an Apple hardware that supports macOS Big Sur 11 which would be >= MBP 2013**

So without new hardware I'm out of luck 😞

**Comment by lukmccall** [comment]
@sshquack, unfortunately, how you said, you won't be able to build the SDK 43 on your current hardware. However, if you want, you can still use dev-client with older SDK (for example 42). Event the newest `expo-dev-client` version (`0.6.3`) should work fine with older SDKs.

@ZeiRashed, if you faced that problem in XCode 13.1, please check the Swift version in your project. If your project is a little bit older and was updated multiple times, the Swift version may not be correct. You can also try to run `pod deintegrate` and then `pod install` to clear everything related to the cocopods.

**Comment by charles-goode** [comment]
Just recently, in a production iOS build, we've been getting in app crashes with
```
EXC_BREAKPOINT
ExpoModulesCore/ConcreteViewProp.swift > Fatal error
``` in the Sentry report. This thread was pretty much the only google result. This occurs when pressing on some iOS notifications. Any thoughts or should I open another issue?

**Comment by lukmccall** [comment]
@charles-goode, please open a new issue for that ;) We will for sure investigate that.

**Comment by tsapeta** [comment]
Hi @charles-goode, do you use `expo-linear-gradient` in your app? So far it's the only module where this error may appear. How do you use it? I mean, what props do you pass there and of what types? Anyway, please answer my questions in a separate issue and mention me there 😉

**Comment by MorganV** [comment]
@tsapeta @lukmccall

I had the same issue on an EAS managed project tracing back to `expo-linear-gradient` (Fatal Error inside `ConcreteViewProp.swift`, but noticed that the same package worked elsewhere in my app. 

It turns out that (at least for me) this code crashes silently:
```
            <LinearGradient
              locations={[0.2, 1]}
              colors={["rgb(30,30,30,)", "rgb(30,30,30)"]}
            >
```

while this does not:
```
            <LinearGradient
              locations={[0.2, 1]}
              colors={["rgb(30,30,30)", "rgb(30,30,30)"]}
            >
```

Note the difference: `"rgb(30,30,30,)"` as opposed to `"rgb(30,30,30)"`.

**Comment by fbruckhoff** [comment]
This isn't working for me either. Have to use the latest Expo SDK because we need support for native custom code due to WebRTC. After upgrading to Xcode 13.1 on MacOS 11.5.2 `expo build:ios` fails consistently even with unmodified template projects, throwing this error:

`⚠️  Something went wrong running pod install in the ios directory.`

Also, I noticed the expo CLI keeps trying to reinstall Cocoapods every time even though it's available.

I thought we can bypass all that by going for EAS Build with a premium plan, but builds are failing there as well.

Would be happy to pay someone help figure this out. Been spending almost a week on this problem now and no luck so far.

**Comment by samwoolertonFosters** [comment]
> 1. `expo init` - tabs template
> 2. cd into project
> 3. `expo install expo-dev-client`
> 4. `expo prebuild`
> 5. `expo run:ios` should work for you here

I followed these steps with Xcode 13 and Expo SDK 43, and got the exact same errors as in the issue description

**Comment by ZeiRashed** [comment]
Just wanted to say I was able to get it working on XCode 13.1. 
In my case, the issue was that I hadn't specified the exact Swift version for one of my targets. When I did that, everything worked!

**Comment by tsapeta** [comment]
Yeah, it's necessary to have Swift version at least 5.4 (shipped as of Xcode 12.5). Closing, but I'll think about better error in that case 😉

**Comment by endruuu** [comment]
Got exactly the same issue. Using EAS build managed workflow on Expo 41

**Comment by tsapeta** [comment]
@endruuu The code that is in the error message was not even available in SDK41 (if you have correct versions of the dependencies). Can you check it using `expo doctor`? Also, next time, please open a new issue when the existing issue was resolved.

**Comment by sbichenko** [comment]
@endruuu Same issue for me.

**Comment by tiavina-mika** [comment]
@ZeiRashed how to do that?

**Comment by tiavina-mika** [comment]
downgrade to sdk 42 fix my issue with macOS Catalina 10.15

**Comment by firofame** [comment]
I'm stuck with Xcode 12.4 due to some dependencies. My project is on react native version 0.61.1. Do I have any options to install expo modules?

**Comment by rajnavadiya6** [comment]
You can install Swift 5.4+ Toolchain inside an older version of XCode. See [these instructions](https://www.hackingwithswift.com/example-code/language/how-to-install-a-beta-version-of-swift).

See all Swift releases [here](https://www.swift.org/download/#releases). I used "Swift 5.5 Development" inside XCode 12.1 on macOS Catalina and that fixed this error for me.

**Comment by yunying1** [comment]
> You can install Swift 5.4+ Toolchain inside an older version of XCode. See [these instructions](https://www.hackingwithswift.com/example-code/language/how-to-install-a-beta-version-of-swift).
> 
> See all Swift releases [here](https://www.swift.org/download/#releases). I used "Swift 5.5 Development" inside XCode 12.1 on macOS Catalina and that fixed this error for me.

Hi, I complete step in the `instructions`, execute `expo run:ios` again, I get the same error, I want to know what steps are missing?

aslo, execute `xcrun swift -version`, output `Apple Swift version 5.1.3 (swiftlang-1100.0.282.1 clang-1100.0.33.15)
Target: x86_64-apple-darwin18.7.0`, I don't know if it's because of this.

**Comment by yunying1** [comment]
I solved this problem through reduce the version of expo, I use SDK 41, run in `xcode` 11.3.1, I have don't change the version of `swift`.
